### ISSUE TITLE
Is there any chance to use Sublime Text's "Symbol List" to outline the stata code?

### ISSUE THREAD

**Issue by zhao2jia** [mattiasnordin/StataEditor#51]
Dear Mattias and other contributors,
Thank you for sharing this plugin. It's a great work!
**I was worderning if there are some ways to get an outline of the whole stata code, especially for GoTo?**
As you might know, if we started a big data project, we might write huge number of code to clean the raw datasets. it's hard to locate the special section. 
To my knowledge, Sublime Text allows users to define local symbol list (Symbol List.tmPreferences), then we can Go To (Ctrl +R) there quickly. is it possible to achieve this?
Many thanks!

**Comment by mattiasnordin** [comment]
Sorry for not having responded in quite some time. I rarely use Stata anymore so not much is happening with the package. Would it be possible for you to create such a symbol list locally or does it have to be included in the package?

**Comment by harningle** [comment]
Yes it's possible. You can find some examples here: https://github.com/harningle/StataEditor/blob/master/Symbol%20List%20-%20Subsubsection.tmPreferences

<image src="https://raw.githubusercontent.com/harningle/StataEditor/docs/docs/assets/images/go_to.gif" width="75%">

**Comment by nicolas-suarez** [comment]
I was coming here to ask about a similar feature. The Stata do-file editor now allows users to define [bookmarks](https://www.stata.com/stata17/do-file-editor/) within the code: you can define a bookmark by typing " **# ". You can also define sublevels within bookmark, so "**##" would be nested one level below an "**##". 

@harningle, do you think this functionality could be integrated into the symbol lists, so we can browse bookmarks within the code in a way that is compatible with vanilla Stata? Here is some sketch code that I wrote (with Copilot's help) to browse bookmarks within a do file. You call this command via the command palette, and then it shows a list of bookmarks, but I feel this "Go To" solution is more elegant .

```
import sublime
import sublime_plugin
import re

class StataBookmarksCommand(sublime_plugin.TextCommand):
    def run(self, edit):
        # Parse bookmarks with hierarchy levels in the current file
        bookmarks = []
        pattern = r"^(\*\*#+)\s*(.+)"  # Match **# level and title
        for region in self.view.find_all(pattern):
            line_number = self.view.rowcol(region.begin())[0] + 1  # Get line number
            match = re.search(pattern, self.view.substr(region))
            if match:
                level = len(match.group(1)) - 2  # Number of # symbols minus the leading **
                title = match.group(2).strip()  # Full title including "bookmark"
                bookmarks.append((level, title, line_number))

        if not bookmarks:
            sublime.message_dialog("No bookmarks found!")
            return

        # Show bookmarks with hierarchy levels in a quick panel
        self.view.window().show_quick_panel(
            [f"{'  ' * level}{title} (Line {line})" for level, title, line in bookmarks],
            lambda index: self.navigate_to(bookmarks, index)
        )

    def navigate_to(self, bookmarks, index):
        # Navigate to the selected bookmark
        if index == -1:  # User canceled
            return
        _, _, line_number = bookmarks[index]
        point = self.view.text_point(line_number - 1, 0)  # Convert line to point
        self.view.sel().clear()
        self.view.sel().add(sublime.Region(point))
        self.view.show(point) 
```

**Comment by harningle** [comment]
> I was coming here to ask about a similar feature. The Stata do-file editor now allows users to define [bookmarks](https://www.stata.com/stata17/do-file-editor/) within the code: you can define a bookmark by typing " **# ". You can also define sublevels within bookmark, so "**##" would be nested one level below an "**##".
> 
> [@harningle](https://github.com/harningle), do you think this functionality could be integrated into the symbol lists, so we can browse bookmarks within the code in a way that is compatible with vanilla Stata? Here is some sketch code that I wrote (with Copilot's help) to browse bookmarks within a do file. You call this command via the command palette, and then it shows a list of bookmarks, but I feel this "Go To" solution is more elegant .

https://github.com/harningle/StataEditor/commit/8a62072a936f1d5aa96b86d9ac27875ec2c4f687 now supports this, but with only up to three nested level, i.e. `**#`, `**##`, `**###`. You can <kbd>ctrl</kbd> + <kbd>R</kbd> to go to symbol list to see the bookmarks.

The reason only three levels are supported is that I don't know sed like regex very well. Not sure how to replace "#" with equal amount of whitespaces to achieve the indent for different level of bookmarks in symbol list. In case anyone wants to have a try on this, my ideal solution would be: match two stars "**", and then match one or more hash keys "#", and a whitespace, and some text. Finally, we add several whitespaces in front to "some text". The number of whitespaces to be added should be in proportion to the number of "#". E.g.

`**# level 1`  --> `level 1` (no whitespace)
`**#### some nested level` --> `   some nested level` (three whitespaces because of four "#")

ps: I left economics and no longer uses Stata anymore. So probably won't be very active maintaining this plugin.
